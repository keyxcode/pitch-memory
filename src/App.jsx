import { useState } from "react";
import Cell, { MiddleCell } from "./components/Cell";
import styled from "styled-components";
import BootstrapContainer from "./components/BootstrapContainer";
import GlobalStyles from "./GlobalStyles";

const StyledBoard = styled.div`
  width: 100%;
  height: 100%;
  display: grid;
  grid-template-columns: ${({ boardSize }) => `repeat(${boardSize}, 1fr)`};
  align-items: center;
  grid-gap: 8px;
`;

function App() {
  const [numCells, setNumCells] = useState(9);
  const [cells, setCells] = useState([
    { selected: false, pitch: "C3", guessed: false },
    { selected: false, pitch: "C4", guessed: false },
    { selected: false, pitch: "C3", guessed: false },
    { selected: false, pitch: "C4", guessed: false },
    { selected: false, pitch: "C3", guessed: false },
    { selected: false, pitch: "C4", guessed: false },
    { selected: false, pitch: "C3", guessed: false },
    { selected: false, pitch: "C4", guessed: false },

    // { selected: false, pitch: "C3", guessed: false },
    // { selected: false, pitch: "C4", guessed: false },
    // { selected: false, pitch: "C3", guessed: false },
    // { selected: false, pitch: "C4", guessed: false },
    // { selected: false, pitch: "C3", guessed: false },
    // { selected: false, pitch: "C4", guessed: false },
    // { selected: false, pitch: "C3", guessed: false },
    // { selected: false, pitch: "C4", guessed: false },
  ]);
  const [message, setMessage] = useState("");

  const handleSelect = (id) => {
    if (cells[id].selected === true) return;

    const newCells = cells.map((cell, i) =>
      i === id ? { ...cell, selected: true } : cell
    );
    setCells(newCells);

    const selectedCells = newCells.filter((cell) => cell.selected === true);
    const numSelectedCells = selectedCells.length;
    console.log("numSelectedCells", numSelectedCells);

    if (numSelectedCells === 1) return;

    if (numSelectedCells === 2) {
      if (guessesAreCorrect(selectedCells)) {
        const cellsMarkedCorrect = markSelectedCellsCorrect(newCells);
        return checkGameOver(cellsMarkedCorrect);
      } else {
        return setAllCellsUnselected();
      }
    }

    return console.log(
      `something wrong happened: num selected = ${numSelectedCells}`
    );
  };

  const guessesAreCorrect = (selectedCells) => {
    if (selectedCells[0].pitch === selectedCells[1].pitch) return true;
    return false;
  };

  const markSelectedCellsCorrect = (cells) => {
    const newCells = cells.map((cell) =>
      cell.selected === true
        ? { ...cell, guessed: true, selected: false }
        : { ...cell, selected: false }
    );

    setTimeout(() => {
      setCells(newCells);
    }, 500);

    return newCells;
  };

  const setAllCellsUnselected = () => {
    const newCells = cells.map((cell) => ({ ...cell, selected: false }));
    return setTimeout(() => {
      setCells(newCells);
    }, 500);
  };

  const checkGameOver = (cells) => {
    const numGuessedCells = cells.filter(
      (cell) => cell.guessed === true
    ).length;
    if (numGuessedCells === cells.length) setMessage("you win");
  };

  const handleRestart = () => {
    const newCells = cells.map((cell) => ({
      ...cell,
      selected: false,
      guessed: false,
    }));
    setCells(newCells);

    setMessage("");
  };

  const boardSize = Math.sqrt(numCells);
  const boardMiddleId = boardSize % 2 === 0 ? null : Math.floor(numCells / 2);

  return (
    <>
      <GlobalStyles />
      <BootstrapContainer>
        {boardMiddleId ? (
          <StyledBoard boardSize={boardSize}>
            {new Array(numCells)
              .fill(0)
              .map((_el, i) =>
                i < 4 ? (
                  <Cell
                    key={i}
                    cellId={i}
                    handleSelect={handleSelect}
                    selected={cells[i].selected}
                    pitch={cells[i].pitch}
                    guessed={cells[i].guessed}
                  />
                ) : i === boardMiddleId ? (
                  <MiddleCell key={i} />
                ) : (
                  <Cell
                    key={i}
                    cellId={i - 1}
                    handleSelect={handleSelect}
                    selected={cells[i - 1].selected}
                    pitch={cells[i - 1].pitch}
                    guessed={cells[i - 1].guessed}
                  />
                )
              )}
          </StyledBoard>
        ) : (
          <StyledBoard>
            {new Array(numCells).fill(0).map((_el, i) => (
              <Cell
                key={i}
                cellId={i}
                handleSelect={handleSelect}
                selected={cells[i].selected}
                pitch={cells[i].pitch}
                guessed={cells[i].guessed}
              />
            ))}
          </StyledBoard>
        )}
        <div>{message && message}</div>
        <button onClick={handleRestart}>Restart</button>
      </BootstrapContainer>
    </>
  );
}

export default App;
