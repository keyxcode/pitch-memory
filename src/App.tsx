import { useMemo, useState } from "react";
import createRandomCells from "./utils/randomCellsGenerator";
import GlobalStyles from "./GlobalStyles";
import MainContainer from "./components/MainContainer";
import ResponsiveContainer from "./components/ResponsiveContainer";
import Board from "./components/Board";
import ButtonGroup from "./components/ButtonGroup";
import Footer from "./components/Footer";
import { Cell } from "./types";

const App = () => {
  const [numCells, setNumCells] = useState(9);
  const [turnsCount, setTurnsCount] = useState(0);
  const [message, setMessage] = useState(
    "find all the squares with the same pitch"
  );
  const [funFact, setFunFact] = useState("");

  const boardSize = useMemo(() => Math.sqrt(numCells), [numCells]);
  const boardMiddleId = useMemo(
    () => (numCells % 2 === 0 ? null : Math.floor(numCells / 2)),
    [numCells]
  );
  const numSoundCells = useMemo(
    () => (numCells % 2 === 0 ? numCells : numCells - 1),
    [numCells]
  );
  const minTurnsCount = useMemo(() => (numSoundCells * 3) / 4, [numSoundCells]);

  const [cells, setCells] = useState(createRandomCells(numSoundCells));

  const handleSelectCell = (id: number): void => {
    if (cells[id].selected === true) return;

    const newCells = cells.map((cell, i) =>
      i === id ? { ...cell, selected: true } : cell
    );
    setCells(newCells);

    const selectedCells = newCells.filter((cell) => cell.selected === true);
    const numSelectedCells = selectedCells.length;

    if (numSelectedCells === 1) return;

    if (numSelectedCells === 2) {
      const newTurnsCount = turnsCount + 1;
      setTurnsCount(newTurnsCount);

      if (guessesAreCorrect(selectedCells)) {
        const cellsMarkedCorrect = markSelectedCellsCorrect(newCells);
        if (gameIsOver(cellsMarkedCorrect)) {
          setMessage(`You win in ${newTurnsCount} turns! ðŸ¥³`);
          setFunFact(
            `Did you know that ${minTurnsCount} turns are the minimum to win this game with no lucky guess?`
          );
        }
      } else {
        setAllCellsUnselected();
      }
    }
  };

  const guessesAreCorrect = (selectedCells: Cell[]): boolean => {
    // console.log(selectedCells);
    if (selectedCells[0].sound === selectedCells[1].sound) return true;
    return false;
  };

  const markSelectedCellsCorrect = (cells: Cell[]): Cell[] => {
    const newCells = cells.map((cell) =>
      cell.selected === true
        ? { ...cell, guessed: true, selected: false }
        : { ...cell, selected: false }
    );

    setTimeout(() => {
      setCells(newCells);
    }, 500);

    return newCells;
  };

  const setAllCellsUnselected = (): void => {
    const newCells = cells.map((cell) => ({ ...cell, selected: false }));
    setTimeout(() => {
      setCells(newCells);
    }, 500);
  };

  const gameIsOver = (cells: Cell[]): boolean => {
    const numGuessedCells = cells.filter(
      (cell) => cell.guessed === true
    ).length;

    if (numGuessedCells === numSoundCells) return true;
    return false
  };

  const handleRestart = (): void => {
    const randomCells = createRandomCells(numSoundCells);
    const newCellsAllSelected = randomCells.map((cell) => ({
      ...cell,
      selected: true,
    }));
    setCells(newCellsAllSelected);

    setMessage("find all the squares with the same pitch");
    setFunFact("");
    setTurnsCount(0);

    setTimeout(() => {
      const newCells = newCellsAllSelected.map((cell) => ({
        ...cell,
        selected: false,
      }));
      setCells(newCells);
    }, 500);
  };

  const handleChangeNumCells = (e: React.ChangeEvent<HTMLSelectElement>) => {
    // console.log("changed board size to", e.target.value);

    const newNumCells = parseInt(e.target.value);
    const newNumSoundCells =
      newNumCells % 2 === 0 ? newNumCells : newNumCells - 1;

    setNumCells(newNumCells);
    setCells(createRandomCells(newNumSoundCells));

    setMessage("find all the squares with the same pitch");
    setFunFact("");
    setTurnsCount(0);
  };

  return (
    <>
      <GlobalStyles />
      <MainContainer>
        <ResponsiveContainer>
          <Board
            boardMiddleId={boardMiddleId}
            boardSize={boardSize}
            numCells={numCells}
            handleSelectCell={handleSelectCell}
            cells={cells}
          />
          <ButtonGroup
            handleRestart={handleRestart}
            handleChangeNumCells={handleChangeNumCells}
            numCells={numCells}
            message={message}
            funFact={funFact}
          />
        </ResponsiveContainer>
        <Footer />
      </MainContainer>
    </>
  );
};

export default App;
