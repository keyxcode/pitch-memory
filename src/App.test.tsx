import { describe } from "vitest";
import { render } from "@testing-library/react";
import App from "./App";

describe("App component", () => {
  it("renders correctly", () => {
    const result = render(<App></App>);
    expect(result).toMatchSnapshot();
  });
});
