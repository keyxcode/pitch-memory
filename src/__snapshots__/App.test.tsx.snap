// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App component > renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bgqQcB UKOcC"
      >
        <div
          class="sc-gTRrQi kAchyq"
        >
          <div
            class="sc-kcuKUB jaVmTh"
          >
            Pitch Memory
          </div>
          <div
            class="sc-dicizt EpAqJ"
          >
            <div
              class="sc-ewnqHT glNfzj"
            >
              <div
                style="display: flex; justify-content: center; align-items: center; height: 100%;"
              />
            </div>
            <div
              class="sc-ewnqHT glNfzj"
            >
              <div
                style="display: flex; justify-content: center; align-items: center; height: 100%;"
              />
            </div>
            <div
              class="sc-ewnqHT glNfzj"
            >
              <div
                style="display: flex; justify-content: center; align-items: center; height: 100%;"
              />
            </div>
            <div
              class="sc-ewnqHT glNfzj"
            >
              <div
                style="display: flex; justify-content: center; align-items: center; height: 100%;"
              />
            </div>
            <div
              class="sc-ewnqHT sc-fFGjHI XnnRw hmQmnO"
            />
            <div
              class="sc-ewnqHT glNfzj"
            >
              <div
                style="display: flex; justify-content: center; align-items: center; height: 100%;"
              />
            </div>
            <div
              class="sc-ewnqHT glNfzj"
            >
              <div
                style="display: flex; justify-content: center; align-items: center; height: 100%;"
              />
            </div>
            <div
              class="sc-ewnqHT glNfzj"
            >
              <div
                style="display: flex; justify-content: center; align-items: center; height: 100%;"
              />
            </div>
            <div
              class="sc-ewnqHT glNfzj"
            >
              <div
                style="display: flex; justify-content: center; align-items: center; height: 100%;"
              />
            </div>
          </div>
          <div
            class="sc-gxYJeL iJggFl"
          >
            <button
              class="sc-iVCKna doCsVB"
            >
              Restart
            </button>
            <select
              class="sc-kPTPQs fBkgRB"
            >
              <option
                value="9"
              >
                3 x 3
              </option>
              <option
                value="16"
              >
                4 x 4
              </option>
              <option
                value="25"
              >
                5 x 5
              </option>
              <option
                value="36"
              >
                6 x 6
              </option>
            </select>
            <div>
              find all the squares with the same pitch
            </div>
          </div>
        </div>
        <div
          class="sc-ispOId czHcJE"
        >
          Copyright © keyxcode
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bgqQcB UKOcC"
    >
      <div
        class="sc-gTRrQi kAchyq"
      >
        <div
          class="sc-kcuKUB jaVmTh"
        >
          Pitch Memory
        </div>
        <div
          class="sc-dicizt EpAqJ"
        >
          <div
            class="sc-ewnqHT glNfzj"
          >
            <div
              style="display: flex; justify-content: center; align-items: center; height: 100%;"
            />
          </div>
          <div
            class="sc-ewnqHT glNfzj"
          >
            <div
              style="display: flex; justify-content: center; align-items: center; height: 100%;"
            />
          </div>
          <div
            class="sc-ewnqHT glNfzj"
          >
            <div
              style="display: flex; justify-content: center; align-items: center; height: 100%;"
            />
          </div>
          <div
            class="sc-ewnqHT glNfzj"
          >
            <div
              style="display: flex; justify-content: center; align-items: center; height: 100%;"
            />
          </div>
          <div
            class="sc-ewnqHT sc-fFGjHI XnnRw hmQmnO"
          />
          <div
            class="sc-ewnqHT glNfzj"
          >
            <div
              style="display: flex; justify-content: center; align-items: center; height: 100%;"
            />
          </div>
          <div
            class="sc-ewnqHT glNfzj"
          >
            <div
              style="display: flex; justify-content: center; align-items: center; height: 100%;"
            />
          </div>
          <div
            class="sc-ewnqHT glNfzj"
          >
            <div
              style="display: flex; justify-content: center; align-items: center; height: 100%;"
            />
          </div>
          <div
            class="sc-ewnqHT glNfzj"
          >
            <div
              style="display: flex; justify-content: center; align-items: center; height: 100%;"
            />
          </div>
        </div>
        <div
          class="sc-gxYJeL iJggFl"
        >
          <button
            class="sc-iVCKna doCsVB"
          >
            Restart
          </button>
          <select
            class="sc-kPTPQs fBkgRB"
          >
            <option
              value="9"
            >
              3 x 3
            </option>
            <option
              value="16"
            >
              4 x 4
            </option>
            <option
              value="25"
            >
              5 x 5
            </option>
            <option
              value="36"
            >
              6 x 6
            </option>
          </select>
          <div>
            find all the squares with the same pitch
          </div>
        </div>
      </div>
      <div
        class="sc-ispOId czHcJE"
      >
        Copyright © keyxcode
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
